arrow-udf-js/src/jsarrow.rs

```diff
@@ -152,116 +110,422 @@ macro_rules! build_array {
+                    }
+                    offsets.push(flatten_values.len() as i32);
+                }
+                let values_array = self.build_array(inner, ctx, flatten_values)?;
+                let nulls = values
+                    .iter()
+                    .map(|v| !v.is_null() && !v.is_undefined())
+                    .collect();
+                Ok(Arc::new(ListArray::new(
+                    inner.clone(),
+                    OffsetBuffer::new(offsets.into()),
+                    values_array,
+                    Some(nulls),
+                )))
+            }
+            DataType::Struct(fields) => {
+                let mut arrays = Vec::with_capacity(fields.len());
+                for field in fields {
+                    let mut field_values = Vec::with_capacity(values.len());
+                    for val in &values {
+                        let v = if val.is_null() || val.is_undefined() {
+                            Value::new_null(ctx.clone())
+                        } else {
+                            let object = val.as_object().context("expect object")?;
+                            object.get(field.name())?
+                        };
+                        field_values.push(v);
+                    }
+                    arrays.push(self.build_array(field, ctx, field_values)?);
                 }
-                arrays.push(build_array(field, ctx, field_values)?);
+                let nulls = values
+                    .iter()
+                    .map(|v| !v.is_null() && !v.is_undefined())
+                    .collect();
+                Ok(Arc::new(StructArray::new(
+                    fields.clone(),
+                    arrays,
+                    Some(nulls),
+                )))
             }
-            let nulls = values
-                .iter()
-                .map(|v| !v.is_null() && !v.is_undefined())
-                .collect();
-            Ok(Arc::new(StructArray::new(
-                fields.clone(),
-                arrays,
-                Some(nulls),
-            )))
+            t => todo!("unsupported data type: {:?}", t),
         }
-        t => todo!("unsupported data type: {:?}", t),
+    }
+
+    fn call_bigdecimal<'a>(
+        &self,
+        ctx: &Ctx<'a>,
+        value: &str,
+    ) -> rquickjs::Result<rquickjs::Value<'a>> {
+        let bigdecimal: Function = ctx.globals().get("BigDecimal")?;
+        bigdecimal.call((value,))
+    }
+
+    fn get_bigdecimal_to_precision_function<'a>(&self, ctx: &Ctx<'a>) -> Result<Function<'a>> {
+        ctx.eval("BigDecimal.prototype.toPrecision")
+            .context("failed to get BigDecimal.prototype.toPrecision")
+    }
+
+    fn decimal_string_to_i128(&self, s: &str, scale: i8) -> Result<i128> {
+        if scale < 0 {
+            return Err(anyhow::anyhow!(
+                "currently only supports non-negative scale"
+            ));
+        }
+
+        let parts = s.split('.').collect::<Vec<&str>>();
```

Comment:
we can assume that there is at most 1 point.

```suggestion
        let (integer, fractional) = match s.split_once('.') {
            Some((i, f)) => (i.parse()..., f[..scale].parse()...),
            None => (s.parse()..., 0)
        };
```

or perhaps we can use the rust-decimal crate to help with it?

Reply:
I was thinking that if we could get by without an added dependency for this use case that would be ideal, but happy to pull in that crate if you prefer!

Reply:
Will get this updated in the next couple of days!